src/drivers/gpio/bcm283x/bcm283x_gpio.c

```diff
@@ -0,0 +1,332 @@
+        REGS_GPIO->re_async_detect_enable.data[port] &= ~pins;
+    } else {
+        REGS_GPIO->re_async_detect_enable.data[port] |= pins;
+    }
+
+    // Async mode = no sampleing, synch = sampleing 
+    // (kind of like a schmitt trigger in that debouncing happens)
+    if(mode & GPIO_MODE_IN_SCHMITT && mode & GPIO_MODE_INT_MODE_FALLING) {
+        REGS_GPIO->fe_async_detect_enable.data[port] &= ~pins;
+    } else {
+        REGS_GPIO->fe_async_detect_enable.data[port] |= pins;
+    }
+
+    // If no interrupt is requested, disable event interrupts
+    //
+    if(!(  mode & GPIO_MODE_INT_MODE_LEVEL1 
+        || mode & GPIO_MODE_INT_MODE_LEVEL0 
+        || mode & GPIO_MODE_INT_MODE_RISING
+        || mode & GPIO_MODE_INT_MODE_FALLING)) {
+        REGS_GPIO->ev_detect_status.data[port] &= ~pins;
+        REGS_GPIO->fe_async_detect_enable.data[port] &= ~pins;
+        REGS_GPIO->re_async_detect_enable.data[port] &= ~pins;
+    }
+*/
+    // Assign pull up / pull down -- floating isn't assigned as it is default state
+    if(mode & (GPIO_MODE_IN_PULL_UP | GPIO_MODE_OUT_PUSH_PULL) ) {
+        bcm293x_gpio_assign_pullup(GPUDUp, port, pins); 
+    }
+    if(mode & (GPIO_MODE_IN_PULL_DOWN | GPIO_MODE_VDD_LEVEL) ) {
+        bcm293x_gpio_assign_pullup(GPUDDown, port, pins); 
+    }
+    
+    // Assign pull up/down/float for all pins in one go (only way allowed)
+    for( GpioPullUpDn i = 0; i < GPIO_PULL_NONE_UP_DN_STATES; i++) {
+        bcm283x_gpio_pin_pullupdown(i, _pinMaskPortA[i], _pinMaskPortB[i]);
+    }
+
+    return 0;
+}
+
+static void bcm283x_gpio_set(unsigned char port, gpio_mask_t pins, char level) {
+    assert(((port == GPIO_PORT_A) || (port == GPIO_PORT_B)));
+    
+    if (level) {
+        REGS_GPIO->output_set.data[port] |= pins;
+    } else {
+        REGS_GPIO->output_clear.data[port] |= pins;
+    }
+}
+
+static gpio_mask_t bcm283x_gpio_get(unsigned char port, gpio_mask_t pins) {
+    assert(((port == GPIO_PORT_A) || (port == GPIO_PORT_B)));
+
+    return REGS_GPIO->level.data[port] & pins;
+}
+
+static struct gpio_chip bcm283x_gpio_chip = {
+    .setup_mode = bcm283x_gpio_setup_mode,
+    .get = bcm283x_gpio_get,
+    .set = bcm283x_gpio_set,
+    .nports = BCM283X_GPIO_PORTS_COUNT
+};
+
+irq_return_t bcm283x_gpio_irq_handler(unsigned int irq_nr, void *data) {
+    uint16_t ports[BCM283X_GPIO_PORTS_COUNT] = {0,0};
+
+    ports[0] = (IRQ_NR_PORTA == irq_nr || IRQ_NR_ALL == irq_nr );
+    ports[1] = (IRQ_NR_PORTB == irq_nr || IRQ_NR_ALL == irq_nr );
+
+    for(uint16_t port = 0; port < BCM283X_GPIO_PORTS_COUNT; port++) {
+        if(ports[port]) {
+            uint32_t pins = REGS_GPIO->ev_detect_status.data[port];
+
+            // Clear the interrupt event on GPIO port
+            REGS_GPIO->ev_detect_status.data[port] = 0xFFFFFFFF;
+
+            gpio_handle_irq(&bcm283x_gpio_chip, port, pins);            
+        }
+    }
+
+    return IRQ_HANDLED;
+}
+
+void bcm283x_gpio_jtag(void) {
```

Comment:
Is it possible to design a separate bcm283x_jtag module (driver) and in the init routine setup necessary pins through global GPIO API?

Reply:
Yes, I suppose it is.  Although, there are only a few viable options available.  In this case, I've made a non-standard selection in effort to keep SPI pins available for other connected devices.  But sure, preferences can be different for other projects.  I'll move it and provide that with 3 configurations where command line can choose another configuration and it will default to one if it is enabled at boot time.